### PR TITLE
MET-1764 Add pending status for consecutive monitor/cancel failures

### PR DESCRIPTION
**Pull request by stzanakis**

### CODE REVIEW

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java [hunk 1]
```diff
@@ -232,35 +247,51 @@ private void periodicCheckingLoop(long sleepTime, AbstractMetisPlugin abstractMe
...
-        LOGGER.warn(String.format("Monitoring of external task failed %s/%s",
-            consecutiveCancelOrMonitorFailures, MAX_CANCEL_OR_MONITOR_FAILURES), e);
-        if (consecutiveCancelOrMonitorFailures == MAX_CANCEL_OR_MONITOR_FAILURES) {
-          monitorResult = new MonitorResult(TaskState.DROPPED, e.getMessage());
-          break;
+        if (ExternalRequestUtil
+            .doesExceptionCauseMatchAnyOfProvidedExceptions(mapWithRetriableExceptions, e)) {
+          consecutiveCancelOrMonitorFailures++;
+          LOGGER.warn(String.format(
+              "Monitoring of external task failed %s. After exceeding %s retries, pending status will be set",
```
Comment (jochen-vermeulen) [inline metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java]: Comment may not be clear to readers... perhaps add the text 'consecutive times' after the first %s?

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java [hunk 2]
```diff
@@ -232,35 +247,51 @@ private void periodicCheckingLoop(long sleepTime, AbstractMetisPlugin abstractMe
...
-        checkPointDateOfProcessedRecordsPeriodInMillis));
+    final boolean notCleaningAndCancelling =
+        abstractMetisPlugin.getPluginStatus() != PluginStatus.CLEANING && workflowExecutionDao
+            .isCancelling(workflowExecution.getId());
+    // A cleaning or a pending task should not be cancelled by exceeding the minute cap
+    final boolean notCleaningOrPendingAndMinuteCapExceeded =
+        (abstractMetisPlugin.getPluginStatus() != PluginStatus.CLEANING
+            && abstractMetisPlugin.getPluginStatus() != PluginStatus.PENDING)
+            && isMinuteCapOverWithoutChangeInProcessedRecords(abstractMetisPlugin,
+            previousProcessedRecords, checkPointDateOfProcessedRecordsPeriodInMillis);
```
Comment (jochen-vermeulen) [inline metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java]: How do you prevent that a task is cancelled by the minute cap as soon as it exits the pending state and becomes running again?

Maybe solve that in isMinuteCapOverWithoutChangeInProcessedRecords() in the same way as with CLEANING?

File: metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/PluginStatus.java [hunk 3]
```diff
@@ -3,21 +3,25 @@
...
 
 /**
+ * The status that a plugin can have.
+ *
  * @author Simon Tzanakis (Simon.Tzanakis@europeana.eu)
  * @since 2017-05-26
  */
 public enum PluginStatus {
-  INQUEUE, CLEANING, RUNNING, FINISHED, CANCELLED, FAILED;
+  INQUEUE, CLEANING, RUNNING, FINISHED, CANCELLED, FAILED, PENDING;
```
Comment (jochen-vermeulen) [inline metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/PluginStatus.java]: I think there are some places where you should treat PENDING the same as RUNNING. I found:
1. WorkflowExecution.setAllRunningAndInqueuePluginsToCancelled()
2. WorkflowExecution.checkAndSetAllRunningAndInqueuePluginsToCancelledIfOnePluginHasFailed()
3. WorkflowExecutor.runInqueueOrRunningStateWorkflowExecution()

Comment (stzanakis) [reply metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/PluginStatus.java]: Indeed, I've fixed that.

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java (hunk 1 above)
Comment (stzanakis) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java]: Fixed

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java (hunk 2 above)
Comment (stzanakis) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java]: Fixed

File: metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/workflow/WorkflowExecution.java [hunk 4]
```diff
@@ -81,25 +81,19 @@ public WorkflowExecution(Dataset dataset, List<AbstractMetisPlugin> metisPlugins
...
+    setAllQualifiedPluginsToCancelled();
     this.setCancelling(false);
   }
 
   /**
    * Checks if one of the plugins has {@link PluginStatus#FAILED} and if yes sets all other plugins
-   * that have status {@link PluginStatus#INQUEUE} or {@link PluginStatus#RUNNING}, to {@link
-   * PluginStatus#CANCELLED}
+   * that have status {@link PluginStatus#INQUEUE} or {@link PluginStatus#RUNNING} or {@link
+   * PluginStatus#CLEANING} or {@link PluginStatus#PENDING, to {@link PluginStatus#CANCELLED}
```
Comment (jochen-vermeulen) [inline metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/workflow/WorkflowExecution.java]: Closing } missing after @link PluginStatus#PENDING in Javadoc.

Comment (jochen-vermeulen) [reply metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/workflow/WorkflowExecution.java]: I fixed it myself.

**Pull request merged**